Haskell multiple outputs: bin etc lib (take three)

domenkozar opened the pull request:
Addresses failures found in https://github.com/NixOS/nixpkgs/pull/32112

Fixes https://github.com/NixOS/nixpkgs/issues/32082

Detailed changelog is in commits, but shortly:

- [x] fixes ghcWithPackages
- [x] hevm
- [x] elm
- [x] ghcjsHEAD 
- [x] agda-prelude 
- [x] happy on darwin
- [x] add tests

@peti I'd kindly ask you for a jobset including darwin, since I can't build all of the packages myself.

cc @nc6

domenkozar committed:
hevm: disable bin output to prevent a cycle

domenkozar committed:
symlinkJoin: support multiple outputs

Matches pkgs.buildEnv behaviour with multiple outputs
and fixes ghcWithPackages using lib as separate output.

domenkozar committed:
elm: fix with separate bin output

domenkozar committed:
ghcjs: fix with separate lib/bin outputs

domenkozar committed:
Agda: fix with separate data output

domenkozar committed:
haskell: generic builder produces working binaries with separate outputs

Sadly on darwin, we always create links file in $lib to workaround
sierra linker limit. This means we have to always create the lib output
even if isLibrary is false.

Moreover, binaries in $bin reference $out even if references files
don't exist. Probably needs fixing in GHC itself. generic-builder
for now just dereferences $bin to $out. This is fragile if a package
uses custom setup.

domenkozar committed:
zsh-git-prompt: fix build for multiple outputs

domenkozar committed:
haskell: add tests covering ghcWith{Packages,Hoogle}

domenkozar committed:
Revert "Merge pull request #32333 from NixOS/revert-32112-hask-split-out"

This reverts commit fd7953d95f478c8cb0ac9db3718adb793142e31d, reversing
changes made to e5e32822df935484f9baff019f692a0c6812f1c4.

domenkozar committed:
elmPackages: statically link haskell deps in executables

Reduces closure size from 2.7GB to ~190MB (where most is python+nodejs)

This optimization is now possible since bin is a separate output
but we can build $lib so that other libraries can depend upon any
elm component.

jb55 reviewed (commented):
Lightly skimmed, untested ACK. Although as a nit the sprinkled getBins seems a bit ugly, but I have no improvements to recommend. :disappointed:

peti commented on pkgs/development/compilers/ghc/8.2.1.nix:
```diff
@@ -10,7 +10,7 @@
   inherit (bootPkgs) ghc;
   version = "8.2.1";
-
+  preReleaseName = "ghc-8.2.1";
```
What is this for?

peti commented on pkgs/development/compilers/halvm/2.4.0.nix:
```diff
@@ -20,9 +20,9 @@ stdenv.mkDerivation rec {
   '';
   configureFlags = stdenv.lib.optional (!enableIntegerSimple) [ "--enable-gmp" ];
-  propagatedNativeBuildInputs = [ alex happy ];
+  propagatedNativeBuildInputs = [ (stdenv.lib.getBin alex) (stdenv.lib.getBin happy) ];
```
My understanding is that a reference like `alex` refers to the output by default that's defined *first*. Is that correct?

If that is so indeed, then we should make sure that the first output defined by Haskell packages is the one containing the binaries so that these references don't need to be wrapped in `getBin` all the time.

domenkozar replied:
`bin` is the first one, this code is just trying to be explicit.

See https://github.com/NixOS/nixpkgs/blob/haskell-outputs-bin-etc-lib/pkgs/development/haskell-modules/generic-builder.nix#L204-L212

peti replied:
I think that being explicit in this case is not an advantage since it makes the code harder to read without adding any functional benefit. People might be confused by those explicit calls, wondering: "Why is this necessary?" Personally, I'd be in favor of dropping `getBin` calls whereever possible. Rather than being explicit, we should document that `bin` can be relied upon to be the first `output` of all Haskell derivations.

dezgeg replied:
It doesn't need to be specifically documented for Haskell since it should hold for all of nixpkgs: https://github.com/NixOS/nixpkgs/blob/master/doc/multiple-output.xml#L39

peti reviewed (commented):
After 2 failed attempts to get these changes merged, I would like someone to actually *use* this code before merging it because verifying the new code by means of Hydra only has clearly failed before. I.e. it would be great if someone would confirm, explicitly, "yes, I have used the code in this PR for the following non-trivial use-cases X and Y for the last 2 weeks and I can confirm that it all works as expected".

domenkozar commented:
@peti I have compiled https://github.com/input-output-hk/cardano-sl with it (non trivial project, 300+ dependencies) and it builds. Haven't deployed it yet.

domenkozar commented:
Created jobset to test all packages at https://hydra.nixos.org/jobset/nixpkgs/pr-32629

grahamc commented:
@GrahamcOfBorg eval

domenkozar commented:
According to https://hydra.nixos.org/eval/1418670?compare=trunk#tabs-new there are following outstanding darwin failures:

- [ ] https://nix-cache.s3.amazonaws.com/log/4qv3l3nzp9zhrqxzymxrn3ia97gnddyw-cabal-helper-0.7.3.0.drv
- [ ] https://nix-cache.s3.amazonaws.com/log/xj3gw998scbjhnpla6wh0g9w95idkw7j-taffybar-with-packages.drv
- [ ] https://nix-cache.s3.amazonaws.com/log/7vlcbyfhf4hbhqkgjcydxy2mym74mrnx-xmonad-with-packages.drv

vcunat commented:
- I have the same xmonad problem on x86_64-linux (as of a0b51de75d).
- When trying that, I noticed that lots of `*-doc` haskell paths are downloaded to build `xmonad-with-packages` &ndash; they don't seem to be big, but it still surprised me.

domenkozar commented:
@vcunat ghcWithPackages installs also `doc` output, I assume that's for having access to docs in the env.

vcunat commented:
Hmm, that might be configurable in `ghcWithPackages`, but it's just a nitpick, and not that much related.

domenkozar commented:
@vcunat fixed :)

domenkozar commented:
I still need to test cabal-helper on darwin, but otherwise this is ready for testing.

domenkozar commented:
@Ericson2314 @shlevy @zimbatm @jb55 @ttuegel @basvandijk @nh2 @nc6

Could I ask you to test this PR with your projects to catch any possible regressions?

bennofs commented on pkgs/development/compilers/elm/default.nix:
```diff
@@ -45,7 +45,7 @@ let
           elmRelease = import ./packages/release.nix { inherit (self) callPackage; };
           elmPkgs' = elmRelease.packages;
-          elmPkgs = elmPkgs' // {
+          elmPkgs = lib.mapAttrs (name: value: hlib.disableSharedExecutables value) (elmPkgs' // {
```
Is this correct? I'm not a user of the elm packages, but this line looks wrong to me: it only disable shared executables for "outside" users of the elm packages, but if any package in the elmPkgs set depends on another one, it will get the shared executable version (b.c. the fixpoint set isn't changed. the modification is applied after the fixpoint).

For example, if we had this situtation:

```
elmPkgs = lib.mapAttrs lib.mapAttrs (name: value: hlib.disableSharedExecutables value) ({
  elm-exe-a = ... derivation depending on self.elm-exe-b
  elm-exe-b = ... some executable ...
})
```

then `elm-exe-a` will get the shared version of `elm-exe-b`, but `elmPkgs.elm-exe-b` will refer to the non-shared version. Maybe it does not matter in this case though because no elm package depends on executables from another one? Or is it actually meant to be only for the "outside"? (why?)

domenkozar replied:
You're right, need to properly override the scope. Couldn't find an easy way to do that without explicitly enumerating all elm packages.

domenkozar replied:
Fixed, this will do a few more rebuilds but it's a proper override.

bennofs commented on pkgs/development/haskell-modules/configuration-nix.nix:
```diff
@@ -217,7 +217,10 @@ self: super: builtins.intersectAttrs super {
-  wxc = (addBuildDepend super.wxc self.split).override { wxGTK = pkgs.wxGTK30; };
+  wxc = (overrideCabal super.wxc (drv: {
+      buildDepends = (drv.buildDepends or []) ++ [self.split];
+      postInstall = "cp -v dist/build/libwxc.so.0.92.3.0 $lib/lib/libwxc.so";
```
Is this related to the multiple-outputs changes or something else? Hardcoding the exact version number here also seems a little brittle. And why is the copy needed, shoudn't cabal install take care to install the wxc library?

bennofs commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -330,18 +357,77 @@ stdenv.mkDerivation ({
+    ${optionalString (hasLibOutput && ! enableSeparateDataOutput) ''
+    # Just like for doc output path in $out potentially landing in
+    # *.conf, we have to also remove the data directory so that it
+    # doesn't appear under data-dir field creating a cycle.
```
this looks to me like it has potential for silently breaking packages: if a library needs access to data files and `enableSeparateDataOutput` is not set, we will brick the library, right? (because references will be replaced with non-existent files). 

Can we limit this to only `.conf` files, and have the build break in other cases?

domenkozar replied:
Good catch.

shlevy commented:
@domenkozar Can you fix conflicts/bring up to date with master and ghc 8.2.2? Then I'll test on our projects.

domenkozar committed:
elmPackages: properly override the whole dep tree

domenkozar committed:
haskell generic builder: remove references only for .conf in libdir

domenkozar commented:
@shlevy done

domenkozar commented:
@GrahamcOfBorg eval

shlevy commented:
@domenkozar:

> $ nix-build -A haskell.packages.ghc802.mtl
> <snip>
> /nix/store/0bvk0df880x9s5sfacb3z97f40lysmff-stdenv/setup: eval: line 1101: unexpected EOF while looking for matching `''
> builder for '/nix/store/25mlqxgzcag7kscqmmbzsqj7w53asb45-mtl-2.2.1.drv' failed with exit code 2
> error: build of '/nix/store/25mlqxgzcag7kscqmmbzsqj7w53asb45-mtl-2.2.1.drv' failed
>

bennofs committed:
haskell generic builder: add missing closing quote

bennofs commented:
@shlevy that was just a missing `'`, I pushed a commit that fixes it

shlevy commented:
> $ nix-build -A haskellPackages.hspec-meta
> these derivations will be built:
>   /nix/store/6sr84x4z0ff9j0xbmm9mfxc43fkx25j4-hspec-meta-2.4.4.drv
> [snip]
> checking for references to /tmp/nix-build-hspec-meta-2.4.4.drv-0 in /nix/store/iji5cysb6cfwzddwh5maz5r31dwak172-hspec-meta-2.4.4-lib...
> cycle detected in the references of '/nix/store/ih60kbmw7wdr7ph9hg77ygn2vn1h017g-hspec-meta-2.4.4' from '/nix/store/iji5cysb6cfwzddwh5maz5r31dwak172-hspec-meta-2.4.4-lib'
>

domenkozar commented:
@shlevy well recent ghc changes broke stuff due to ghc upgrade ..

shlevy commented:
Sure just reporting results of my testing :)

domenkozar commented:
I'll block improvements here until https://github.com/NixOS/nixpkgs/pull/33054 lands and upcoming changes for faster generic builder iterations.

domenkozar commented:
Binary GHC for bootstrapping landed in master, I've rebased this and a bunch of things fail. I'll report back when it's all in good state again.

vcunat commented:
_Disabled notification on the jobset, after getting lots of "success" e-mails._

copumpkin commented:
Holy **** yes my inbox got flooded, thanks for disabling that 😄

domenkozar commented:
Sorry about that - I copied trunk jobset and forgot to switch off notifications.

roberth commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -330,18 +357,77 @@ stdenv.mkDerivation ({
+    ${optionalString enableSeparateBinOutput "mkdir -p ${binDir} ${libexecDir}"}
+    ${optionalString enableSeparateEtcOutput "mkdir -p ${etcDir}"}
+
+    ${optionalString (stdenv.isDarwin && enableSeparateBinOutput && ! hasLibOutput) ''
```
Is there a particular reason to do this on darwin only?

`happy` seems to [need](https://hydra.nixos.org/build/66728685/nixlog/1) something like this on linux.

Why not something like:

    configureFlags = drv.configureFlags or [] ++ optional (enableSeparateBinOutput && ! hasLibOutput) "--libdir=/var/empty";

That way we don't have to `remove-references-to` and it may (?) fail earlier. I'd rather replace it with an empty directory in the nix store though.

roberth committed:
haskellPackages.happy: fix circular reference caused by separate bin

roberth committed:
haskell generic builder: make robust against missing lib directory

Darwin builds always have a lib *output* but may not produce a lib
*directory* inside of it when they are not library packages and
the executable does not require dylibs for circumventing Sierra
linker limits.

roberth committed:
haskellPackages.hspec-{meta,discover}: resolve store reference cycles

roberth committed:
haskellPackages.uuagc-bootstrap: don't haddock

It is just uuagc itself but preprocessed by uuagc.
Not haddocking should save some space and build time.

roberth committed:
haskell generic builder: broaden datadir removal from lib

roberth committed:
haskell generic builder: avoid redundant lib output on darwin

The purpose here is to address the failure mode where $lib is not
created. Additionally this reduces the number of store paths and
nars.

roberth committed:
haskell generic builder: use setup dependency libs

ryantm commented:
They are many references to this PR fixing open Haskell issues. What's the status of it?

domenkozar commented:
It's kind of broken until staging is merged, then we can asses it's status.

roberth commented:
New todo:

 - [ ] implement 'safe' `remove-references` that checks whether either target is empty or reference wasn't made
 - [ ] simplify by applying the safe variation generously
 - [ ] rebase
 - [ ] test using hydra job for this branch

dezgeg commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -222,7 +253,15 @@ assert allPkgconfigDepends != [] -> pkgconfig != null;
   name = "${pname}-${version}";
 
-  outputs = [ "out" ] ++ (optional enableSeparateDataOutput "data") ++ (optional enableSeparateDocOutput "doc");
+  outputs =
```
Hmm, no `dev` output? I would expect one to contain all the `.o` and `.hi` files, plus the GHC package `.conf` file.

dezgeg commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -246,14 +285,14 @@ stdenv.mkDerivation ({
   postPatch = optionalString jailbreak ''
     echo "Run jailbreak-cabal to lift version restrictions on build inputs."
-    ${jailbreak-cabal}/bin/jailbreak-cabal ${pname}.cabal
+    ${stdenv.lib.getBin jailbreak-cabal}/bin/jailbreak-cabal ${pname}.cabal
```
It's not needed to call the `getBin` here. The rule of multiple outputs is that `${foo}/bin/bar` will always, always do the right thing.

nh2 commented:
Can this also split out profiling builds? That would be useful.

domenkozar commented:
I think if we scale down ambitions of this PR and just get optional `bin` output first, this would be a small step forward. Enabling it by default can be phase 2. Same goes for the rest of splits.

domenkozar commented:
Seems like that's the new plan then.

domenkozar commented:
https://github.com/NixOS/nixpkgs/pull/58525
